lgc/builder/BuilderBase.cpp

```diff
@@ -30,11 +30,25 @@
  */
 #include "lgc/util/BuilderBase.h"
 #include "llvm/IR/InlineAsm.h"
+#include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/IntrinsicsAMDGPU.h"
 
 using namespace lgc;
 using namespace llvm;
 
+// =====================================================================================================================
+// Set the insert point to be just past the initial block of allocas in the given function's entry block.
+//
+// Use this method if you need to insert code to define values that are accessible in the entire function.
+void BuilderCommon::setInsertPointPastAllocas(Function &fn) {
+  BasicBlock &bb = fn.getEntryBlock();
+  auto it = bb.begin();
+  while (it != bb.end() &&
+         (isa<PHINode>(*it) || isa<AllocaInst>(*it) || isa<DbgInfoIntrinsic>(*it) || it->isLifetimeStartOrEnd()))
+    ++it;
+  SetInsertPoint(&bb, it);
+}
```

Comment:
https://llvm.org/docs/LangRef.html#functions

> The first basic block in a function is special in two ways: it is immediately executed on entrance to the function, and it is not allowed to have predecessor basic blocks (i.e. there can not be any branches to the entry block of a function). Because the block can have no predecessors, it also cannot have any [PHI nodes](https://llvm.org/docs/LangRef.html#i-phi).

Reply:
Right, phis don't need to be tested. I left out `pseudoprobe` because I don't see us using it anytime soon. It isn't even documented properly. Lifetime markers are less clear to me but probably we shouldn't test for those either.